### PR TITLE
Impl recorder part

### LINKED ISSUE #361: Implement LiquidStaking2.0, stake-client interaction part
**Motivation**

On-chain staking pallet needs to interact with off-chain stake-client, here are the main methods.
Please adjust if necessary.

let's develop from this pr, #362 @alannotnerd 

**Suggested Solution**

- [x] trigger_new_era
- [x] record_reward
- [x] record_slash
- [x] record_bond_response/record_bond_extra_response/record_rebond_response/record_unbond_response
- [x] transfer_to_relaychain

### PR DESCRIPTION
**Pull request by alannotnerd**
WIP: Impl recorder part,

Close #361

### CODE REVIEW

File: pallets/liquid-staking-two/src/lib.rs [hunk 1]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
             origin: OriginFor<T>,
             era_index: EraIndex,
         ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?; //TODO(Alan): Who can call this extrinsic?
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:252]: We can refer to this:
https://github.com/parallel-finance/parallel/blob/master/pallets/liquid-staking/src/lib.rs#L347

File: pallets/liquid-staking-two/src/lib.rs [hunk 2]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
...
+            CurrentEra::<T>::put(era_index);
+            // TODO(Alan): Should we deposit event?
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:254]: I think an event is necessary

File: pallets/liquid-staking-two/src/lib.rs [hunk 3]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
...
-        #[transactional]
-        pub fn record_bond_extra_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub fn record_reward(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            #[pallet::compact] amount: Balance,
+            block_number: BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            let _sender = ensure_signed(origin)?;
+            MatchingPool::<T>::try_mutate(
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:273]: `MatchingPool` mainly works when user stake and unstake, please refer to :
https://github.com/parallel-finance/parallel/blob/master/pallets/liquid-staking/src/lib.rs#L365

when record_reward or record_slash, exchangerate will be updated.

excahngerate = stakingcurrency amount/ liquidcurrency amount；
the amount of stakingcurrency(KSM) will be store in `StakingPool`,
the amount of liquidcurrency(xKSM) can get like this: T::Currency::free_balance()

File: pallets/liquid-staking-two/src/lib.rs [hunk 4]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
...
-        #[pallet::weight(10_000)]
-        #[transactional]
-        pub fn record_bond_extra_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub fn record_reward(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            #[pallet::compact] amount: Balance,
+            block_number: BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            let _sender = ensure_signed(origin)?;
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:272]: here we also need to confirm if it's an approved sender.

File: pallets/liquid-staking-two/src/lib.rs [hunk 5]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
...
+                        .ok_or(ArithmeticError::Overflow)?;
+                    *matching_buffer = MatchingBuffer {
+                        total_stake_amount,
+                        ..matching_buffer.clone()
+                    };
+                    Ok(().into())
+                },
+            )?;
+
+            StakingOperationHistory::<T>::insert(
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:288]: in every era, we can only have one record_reward operation, so I think it's better to check if it already exists.

and then, we insert it into operation history.

File: pallets/liquid-staking-two/src/lib.rs [hunk 6]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
...
 
+        /// Invoked by offchain stake client. Declare reward received.
+        ///
+        /// `origin` should be approved account.(not implement yet).
+        ///
+        /// Q: Should update exchange_rate?
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn record_rebond_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub fn record_slash(
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:307]: record_slash is similar to record_reward.

File: pallets/liquid-staking-two/src/lib.rs [hunk 7]
```diff
@@ -247,45 +249,116 @@ pub mod pallet {
...
             Ok(().into())
         }
 
+        /// Invoked by offchain stake client.
+        ///
+        /// Record bond operation and corresponding data.
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn record_unbond_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub fn record_bond_response(
```
Comment (MrPai) [inline pallets/liquid-staking-two/src/lib.rs:346]: for this method, I think we need first to check if the `StakingOperationType` exists and its status is not successful,

**Review by MrPai — COMMENTED** [review]
I left some comments, maybe you can refer to them. let's work on it together. @alannotnerd

File: pallets/liquid-staking-two/src/lib.rs (hunk 7 above)
Comment (alannotnerd) [reply pallets/liquid-staking-two/src/lib.rs:346]: Could you please explain? I'm kind of confused about it. Should there be only one bonding operation in an era?

File: pallets/liquid-staking-two/src/lib.rs (hunk 3 above)
Comment (alannotnerd) [reply pallets/liquid-staking-two/src/lib.rs:273]: free_balance of whom? The agent?

File: pallets/liquid-staking-two/src/lib.rs [hunk 8]
```diff
@@ -146,6 +146,12 @@ pub mod pallet {
     pub enum Error<T> {
         /// ExchangeRate is invalid
         InvalidExchangeRate,
+        /// Operation is not allowed
+        OperationNotAllowed,
+        /// Caller is not approved.
+        IllegalAgent,
+        /// Slash or reword event is recored.
```
Comment (ashi20192020ayush) [inline pallets/liquid-staking-two/src/lib.rs:153]: Please correct it `Slash or record event is recorded.` 🙂

File: pallets/liquid-staking-two/src/lib.rs [hunk 9]
```diff
@@ -146,6 +146,12 @@ pub mod pallet {
...
+        EventRecorded,
```
Comment (ashi20192020ayush) [inline pallets/liquid-staking-two/src/lib.rs:154]: Can we rename it? It doesn't look like error message. @yz89 What do you suggest?

File: pallets/liquid-staking-two/src/lib.rs [hunk 10]
```diff
@@ -161,6 +167,10 @@ pub mod pallet {
         RewardsRecorded(T::AccountId, Balance),
         /// The slash is recorded
         SlashRecorded(T::AccountId, Balance),
+        /// Era advanced.
```
Comment (ashi20192020ayush) [inline pallets/liquid-staking-two/src/lib.rs:170]: Can we make it consistent? It should be `Era Index is updated.`

File: pallets/liquid-staking-two/src/lib.rs [hunk 11]
```diff
@@ -161,6 +167,10 @@ pub mod pallet {
...
+        EraIndexUpdated(EraIndex),
+        /// A staking operation reponse recorded.
+        ResponseRecorded(StakingOperationType, Balance, ResponseStatus, BlockNumber),
```
Comment (ashi20192020ayush) [inline pallets/liquid-staking-two/src/lib.rs:173]: Similarly `A staking operation response is recorded.`

File: pallets/liquid-staking-two/src/lib.rs [hunk 12]
```diff
@@ -245,47 +255,141 @@ pub mod pallet {
...
-        pub fn record_bond_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            Ok(().into())
-        }
-
-        #[pallet::weight(10_000)]
-        #[transactional]
-        pub fn record_bond_extra_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            Ok(().into())
-        }
+        pub fn record_slash(
```
Comment (ashi20192020ayush) [inline pallets/liquid-staking-two/src/lib.rs:325]: This looks similar to `record_reward`. Can we make this function generic to avoid duplicate code.

**Review by ashi20192020ayush — COMMENTED** [review]
It seems this PR is WIP. Please add corresponding test-cases.

File: pallets/liquid-staking-two/src/lib.rs (hunk 7 above)
Comment (MrPai) [reply pallets/liquid-staking-two/src/lib.rs:382]: yes, each era can only have one bonding or unbonding operation

File: pallets/liquid-staking-two/src/lib.rs (hunk 3 above)
Comment (MrPai) [reply pallets/liquid-staking-two/src/lib.rs]: should be T::Currency::total_issuance()

**Comment by yz89** [comment]
@ashi20192020ayush Thank you for the comments.

**Pull request merged**